### PR TITLE
[bzip2] Initial implementation

### PR DESCRIPTION
**Pull request by oleid**
This is an attempt to add bzip2 support to this library. It's at RFC stage, since a few unreleased  bzip2 changes are still needed, c.f. https://github.com/alexcrichton/bzip2-rs/pull/49. 

EDIT: It now works with the latest stable version

## Description
Since flate and bzip2 crates are quite similar, I used that as a basis for the implementation. Only details are different.

## How Has This Been Tested?
I added a test case. It still fails, but I'm not entirely sure why. Probably something is wrong at compression, since I can decompress just fine using the crate in a (still private) app. Maybe you have some more insight? 

EDIT: Tests pass fine :).

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the [CONTRIBUTING](https://github.com/rustasync/tide/blob/master/.github/CONTRIBUTING.md) document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by Nemo157** [comment]
The tests are quite aggressive about using small input and output chunk sizes (only 2 bytes in a lot of cases) to try and trigger edge cases, so it's quite likely that it can work in a real usecase but fail the tests.

Looking at what happens it looks like bzip2 decompression doesn't distinguish between end of input and completion of flushing output, so making `BzDecoder::finish` just return `Ok(true)` passes all of the tests.

For compression I think `FinishOk` just means it successfully flushed some output data, but there is still data remaining to flush, once it's complete it returns `StreamEnd`. Swapping the return values for those statuses got all the compression tests passing.

You can see the debugging I added to find that, and the changes that made it pass here: <https://github.com/Nemo157/async-compression-rs/commit/f009b68ea02225bbf51dc6b7049979aad3de7532>

File: src/codec/bzip2/decoder.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            // The stream's end has been met, meaning that no more data can be input.
+            Status::StreamEnd => Ok(true),
+
+            // There was insufficient memory in the input or output buffer to complete
+            // the request, but otherwise everything went normally.
+            Status::MemNeeded => Err(Error::new(ErrorKind::Other, "out of memory")),
+        }
+    }
+
+    fn finish(&mut self, output: &mut PartialBuffer<&mut [u8]>) -> Result<bool> {
```
Comment (Nemo157) [inline src/codec/bzip2/decoder.rs]: So this should just return `Ok(true)` since it was already finished during the `decode`.

File: src/codec/bzip2/encoder.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            Status::Ok => Ok(false),
+
+            // The Flush action on a compression went ok.
+            Status::FlushOk => unreachable!(),
+
+            // The Run action on compression went ok.
+            Status::RunOk => unreachable!(),
+
+            // The Finish action on compression went ok.
+            Status::FinishOk => Ok(true),
```
Comment (Nemo157) [inline src/codec/bzip2/encoder.rs]: This should be `Ok(false)`

File: src/codec/bzip2/encoder.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+            // The stream's end has been met, meaning that no more data can be input.
+            Status::StreamEnd => unreachable!(),
```
Comment (Nemo157) [inline src/codec/bzip2/encoder.rs]: And this should be `Ok(true)`

**Comment by oleid** [comment]
Thanks for taking the time to debug the issue!

I guess we could manually convert the error for now; not sure why I haven't done that in the first place ;)

Anyway, would you like me to clean up the commits some more (I kept them separate for further review), or shall we just squash them together during merge?

**Comment by Nemo157** [comment]
Thanks for implementing this!

Feel free to squash then `bors r+` when it's ready.

bors delegate+

**Comment by Nemo157** [comment]
Oh, just remembered about proptest tests as well, it'd be great if you could add `bzip2` to the algos tested here as well

https://github.com/rustasync/async-compression/blob/c5d9bcb0d3a87aaaa034370cc7d1418c67bc52ee/tests/proptest.rs#L61

**Comment by oleid** [comment]
> Oh, just remembered about proptest tests as well, it'd be great if you could add `bzip2` to the algos tested here as well
> 
> https://github.com/rustasync/async-compression/blob/c5d9bcb0d3a87aaaa034370cc7d1418c67bc52ee/tests/proptest.rs#L61

Done!

**Comment by oleid** [comment]
bors +r

**Comment by oleid** [comment]
> Did you mean "r+"?

Yes, I did!

bors "r+"

**Comment by oleid** [comment]
Oh brave new world, that has such wonderful bots in it! :smile:
